stablehlo/dialect/TypeInference.cpp

```diff
@@ -574,60 +574,16 @@ LogicalResult inferConcatenateOp(Optional<Location> location, ValueRange inputs,
   auto firstType = inputs[0].getType().cast<ShapedType>();
   auto outElement = firstType.getElementType();
 
-  // Find the first ranked input to determine the output rank.
-  for (auto type : inputs.getTypes()) {
-    auto shapedType = type.cast<ShapedType>();
-    if (shapedType.hasRank()) {
-      firstType = shapedType;
-      break;
-    }
-  }
-
-  // If all inputs are unranked, the result must be unranked.
-  if (!firstType.hasRank()) {
-    inferredReturnTypes.push_back(UnrankedTensorType::get(outElement));
-    return success();
-  }
-
-  auto outShape = llvm::to_vector<6>(firstType.getShape());
-
-  // Determine what the non-concatenate dimensions should be.
-  for (auto type : inputs.getTypes()) {
-    auto shapedTy = type.cast<ShapedType>();
-    if (!shapedTy.hasRank()) {
-      continue;
-    }
-
-    for (const auto& it : llvm::enumerate(shapedTy.getShape())) {
-      // If a dimension is not dynamic, the output shape should match.
-      if (ShapedType::isDynamic(outShape[it.index()])) {
-        outShape[it.index()] = it.value();
-      }
-    }
-  }
-
-  outShape[dimension] = 0;
-
   for (auto operand : inputs.getTypes()) {
```

Comment:
isn't this conservative compared to binary op?

In case of inputs tensor<2x?xf32> and tensor<*xf32> with dim = 1, we know that the result type for legal cases would be tensor<2x?xf32>.

Reply:
You example is very good: inputs: 
<2x?xf32>, <*xf32> with concat dim 1 should infer <2x?xf32>
This can be addressed by ingoreing all the unranked inputs and only the the ranked ones.
But soon I realize the above claim is wrong: for example in case of
<2x?xf32>, <*xf32> with concat dim 0 should infer <?x?xf32> instead of <2x?xf32>
My solution to all these are: treat unranked input type as a fully dynamic type <?x...?x>. Please see the new commit.f